sdks/js-sdk/README.md

```diff
@@ -0,0 +1,150 @@
+
+## Adding a module to the JS SDK
+
+New modules written JavaScript or Typescript will have the following items:
+
+- A single `index` entry file that organizes and exposes all the exported contents of the module.
+- Manually written or compiled files for a Cordova plugin should be present in the `www/` folder by convention
+- Native code in the specific platform folder. i.e. `android/`, `ios/`
+- A Cordova `plugin.xml` that references the code contents above via `<js-module>` and `<platform>` tags
+
+### Adding native bindings to a plugin for Cordova
+
+Native bindings can be required for some specific features of a 'main' JavaScript module.
+These should be exposed as a separate NPM-installable plugin that expose the required features by the 'main' module.
+
+The main module then feature-checks for the existance of the native feature and deal accordingly, allowing for a single codebase to adapt to different
+execution environments and for graceful degradation.
+
+The following diagram shows the relationship between native plugins, JavaScript code and client App code:
+
+<!-- TODO: include diagram -->
+
+#### `plugin.xml` example
+
+```xml
+<platform name="android">
+  <config-file target="res/xml/config.xml" parent="/*">
+      <feature name="PlatformVersionModule" >
+          <param
+              name="android-package"
+              value="org.aerogear.mobile.core.PlatformVersionModule"
+          />
+      </feature>
+  </config-file>
+
+  <js-module>
+    <clobbers target="aerogear.platformVersion">
+  </js-module>
+
+  <config-file target="AndroidManifest.xml" parent="/*">
+      <uses-permission android:name="android.permission.INTERNET" />
+  </config-file>
+
+  <source-file
+      src="src/android/PlatformVersionModule.java"
+      target-dir="src/org/aerogear/mobile/core/PlatformVersionModule"
+  />
+</platform>
+```
+
+#### Native plugin usage example
+
+```javascript
+// check for plugin presence
+if (aerogear.platformVersion) {
+    // utilize feature
+    const version = ;
+    aerogear.platforVersion.get()
+        .then(version => metricsService.sendAppMetrics({
+            //...
+            platforVersion: version
+        }));
+} else {
+    // take appropriate measures
+}
+```
+
+### Adding native bindings to a plugin for React Native
+
+Native bindings for React Native should also be delivered as a separate NPM package and register their native modules in react-native's `NativeModules` object.
+The code in the 'main' JavaScript module should be able to check for the module's presence in order to react accordingly.
+
+Optionally these can include typing declarations or convenience JavaScript to facilitate access, but these should still allow for feature-checking by the main module.
+
+## Investigation
+
+A Proof of Concept was done in the following repository: https://github.com/aerogear/aerogear-js-sdk
+
+### PoC results for the Core Module for the JS SDK
```

Comment:
This kinda shows the thinking progress. We can revert that if needed by let's remove it!